{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from collections import defaultdict\n",
    "from copy import deepcopy\n",
    "import networkx as nx \n",
    "import matplotlib.pyplot as plt \n",
    "import sys\n",
    "from pyvis.network import Network\n",
    "\n",
    "sys.setrecursionlimit(10000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "np.random.seed(0)\n",
    "\n",
    "nucleotides = ['A', 'T', 'G', 'C']\n",
    "\n",
    "sequence_string = ''.join(np.random.choice(nucleotides, 1000).tolist())\n",
    "reads = ''\n",
    "\n",
    "for i in range(1, 1001):\n",
    "    read_len = 150\n",
    "    read_start = np.random.default_rng().integers(0, 1000 - read_len)\n",
    "    read_str = sequence_string[read_start:read_start+read_len]\n",
    "    reads += f'>READ_SEQ_{i}\\n{read_str}\\n'\n",
    "\n",
    "with open(f\"reads.fasta\", \"w\") as file:\n",
    "    file.write(reads)\n",
    "\n",
    "with open(\"sequense.fasta\", \"w\") as file:\n",
    "    file.write('>MAIN_SEQ_0\\n')\n",
    "    file.write(sequence_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Построение графа Де Брюина\n",
    "По заданному набору ридов в формате FASTQ и параметру k, который соответствует длине k-меров, построить граф Де Брюина, некоторый путь в котором соответствовал бы возможной подстроке в исходном геноме. Не забывайте про запоминание покрытия каждого k-мера, а так же про сами подстроки, которые соответствуют каждому ребру. В остальном граф полностью соответствует тому, что был описан в лекции."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_name, seq_string = [], []\n",
    "with open (\"/home/pk/Desktop/BOTAY/BioinformaticsCourse2024PK/homework/2_2/reads.fasta\",'r') as fa:\n",
    "    for seq_record  in SeqIO.parse(fa,'fasta'):\n",
    "        seq_name.append(str(seq_record.name))\n",
    "        seq_string.append(str(seq_record.seq))\n",
    "\n",
    "true_seq_name, true_seq_string = [], []\n",
    "with open (\"/home/pk/Desktop/BOTAY/BioinformaticsCourse2024PK/homework/2_2/sequense.fasta\",'r') as fa:\n",
    "    for seq_record  in SeqIO.parse(fa,'fasta'):\n",
    "        true_seq_name.append(str(seq_record.name))\n",
    "        true_seq_string.append(str(seq_record.seq))\n",
    "true_seq = true_seq_string[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test.html\n"
     ]
    }
   ],
   "source": [
    "class Node():\n",
    "    def __init__(self, string):\n",
    "        self.string = string\n",
    "        self.in_degree = 0\n",
    "        self.out_degree = 0\n",
    "\n",
    "class Edge():\n",
    "    def __init__(self, prev_node, next_node, string, coverage=1):\n",
    "        self.prev_node = prev_node\n",
    "        self.next_node = next_node\n",
    "        self.string = string\n",
    "        self.coverage = coverage\n",
    "\n",
    "class DBGraph():\n",
    "    def __init__(self, reads, k):\n",
    "        self.reads = reads\n",
    "        self.k = k\n",
    "        self.used_edges = set()\n",
    "        self.used_nodes = set()\n",
    "        self.nodes = {}\n",
    "        self.edges = {}\n",
    "        self.connections = defaultdict(list)\n",
    "        self.kmers = []\n",
    "        self.compacted = False\n",
    "        self.deleted_tails = False\n",
    "        self.string = None\n",
    "\n",
    "    def gen_kmers(self):\n",
    "        for read in self.reads:\n",
    "            self.kmers.extend([read[i:self.k+i] for i in range(0, len(read)-self.k+1)])\n",
    "\n",
    "    def create_graph(self):\n",
    "        for kmer in self.kmers:\n",
    "            prev_node_str = kmer[:-1]\n",
    "            next_node_str = kmer[1:]\n",
    "            if prev_node_str not in self.used_nodes:\n",
    "                self.nodes[prev_node_str] = Node(prev_node_str)\n",
    "                self.used_nodes.add(prev_node_str)\n",
    "            if next_node_str not in self.used_nodes:\n",
    "                self.nodes[next_node_str] = Node(next_node_str)\n",
    "                self.used_nodes.add(next_node_str)\n",
    "            if kmer in self.used_edges:\n",
    "                self.edges[kmer].coverage += 1\n",
    "            else:\n",
    "                self.used_edges.add(kmer)\n",
    "                self.edges[kmer] = Edge(prev_node_str, next_node_str, kmer)\n",
    "                self.connections[prev_node_str].append(self.edges[kmer])\n",
    "                self.nodes[prev_node_str].out_degree += 1\n",
    "                self.nodes[next_node_str].in_degree += 1\n",
    "\n",
    "    def compact(self):\n",
    "        # if self.compacted:\n",
    "        #     print('This graph has already been compacted')\n",
    "        #     return\n",
    "        to_delete = set()\n",
    "        self.compacted = True\n",
    "        self.connections_temp = deepcopy(self.connections)\n",
    "        for node in self.nodes:\n",
    "            # print(node)\n",
    "            # print(self.nodes[node].in_degree)\n",
    "            # print(self.nodes[node].out_degree)\n",
    "            if self.nodes[node].out_degree >= 1:\n",
    "                for new_edge_idx in range(len(self.connections[node])):\n",
    "                    new_edge = self.connections[node][new_edge_idx]\n",
    "                    next_edge = new_edge\n",
    "                    # print(next_edge.next_node)\n",
    "                    if self.nodes[new_edge.next_node].out_degree == 1 and self.nodes[new_edge.next_node].in_degree == 1:\n",
    "                        while True:\n",
    "                            prev_edge = new_edge\n",
    "                            next_edge = self.connections[new_edge.next_node][0]\n",
    "                            new_edge_string = prev_edge.string + next_edge.string[-1]\n",
    "                            new_edge_coverage = int(len(prev_edge.string)/len(new_edge_string)*prev_edge.coverage + len(next_edge.string)/len(new_edge_string)*next_edge.coverage)\n",
    "                            # self.nodes[prev_edge.next_node].is_compacted = True\n",
    "                            self.nodes[prev_edge.next_node].in_degree -= 1\n",
    "                            self.nodes[prev_edge.next_node].out_degree -= 1\n",
    "                            del self.connections[prev_edge.next_node]\n",
    "                            new_edge = Edge(node, next_edge.next_node, new_edge_string, new_edge_coverage)\n",
    "                            if self.nodes[new_edge.next_node].out_degree != 1 or self.nodes[new_edge.next_node].in_degree != 1:\n",
    "                                # del self.connections[new_edge.prev_node]\n",
    "                                break\n",
    "                    self.connections[node][new_edge_idx] = new_edge\n",
    "        to_delete = set(self.nodes)\n",
    "        for node in self.connections:\n",
    "            self.nodes[node].out_degree = len(self.connections[node])\n",
    "            for edge in self.connections[node]:\n",
    "                to_delete.discard(edge.prev_node)\n",
    "                to_delete.discard(edge.next_node)\n",
    "        for node in list(to_delete):\n",
    "            del self.nodes[node]\n",
    "\n",
    "    def _find_tails(self, v):\n",
    "        for edge in self.connections_temp[v]:\n",
    "            if edge.coverage:\n",
    "                edge.coverage -= 1\n",
    "                if self.nodes[edge.next_node].out_degree <= 1 and edge.prev_node not in self.nodes_in_tail:\n",
    "                    prev_start_node = edge.prev_node\n",
    "                    start_node = edge.next_node\n",
    "                    edge_temp = edge\n",
    "                    tail_coverage = 0\n",
    "                    tail_len = 0\n",
    "                    while True:\n",
    "                        if self.nodes[edge_temp.next_node].out_degree > 1:\n",
    "                            break\n",
    "                        tail_len += 1\n",
    "                        tail_coverage += edge_temp.coverage\n",
    "                        if self.nodes[edge_temp.next_node].out_degree == 0:\n",
    "                            self.tails[(prev_start_node, start_node)] = tail_len*tail_coverage\n",
    "                            break\n",
    "                        edge_temp = self.connections[edge_temp.next_node][0]\n",
    "                        self.nodes_in_tail.add(edge_temp.prev_node)\n",
    "                self._find_tails(edge.next_node)\n",
    "\n",
    "    def _check_tails(self):\n",
    "        for node in self.nodes:\n",
    "            if self.nodes[node].out_degree == 0:\n",
    "                return True\n",
    "        return False\n",
    "\n",
    "    def delete_tails(self):\n",
    "        if not self._check_tails():\n",
    "            print('This graph does not have tails!')\n",
    "            return\n",
    "        # if self.deleted_tails:\n",
    "        #     print('Tails have already been deleted')\n",
    "        #     return\n",
    "        # self.delete_tails = True\n",
    "        self.nodes_in_tail = set()\n",
    "        self.tails = {}\n",
    "        in_degrees = [self.nodes[i].in_degree for i in self.connections.keys()]\n",
    "        nodes_min_in_degrees = [list(self.connections.keys())[i] for i in range(len(in_degrees)) if in_degrees[i] == min(in_degrees)]\n",
    "        self.connections_temp = deepcopy(self.connections)\n",
    "        for start in nodes_min_in_degrees:\n",
    "            self._find_tails(start)\n",
    "        if len(self.tails) > 1:\n",
    "            tails_min_scores = set(np.sort(list(self.tails.values()))[:int(len(self.tails)*0.3)+1])\n",
    "            for tail in self.tails:\n",
    "                if self.tails[tail] in tails_min_scores:\n",
    "                    self.connections[tail[0]] = [edge for edge in self.connections[tail[0]] if edge.next_node != tail[1]]\n",
    "                    self.nodes[tail[0]].out_degree = len(self.connections[tail[0]])\n",
    "                    edge = self.connections[tail[1]][0]\n",
    "                    while True:\n",
    "                        del self.connections[edge.prev_node]\n",
    "                        del self.nodes[edge.prev_node]\n",
    "                        if not self.connections[edge.next_node]:\n",
    "                            break\n",
    "                        edge = self.connections_temp[edge.next_node][0]\n",
    "                    del self.connections[edge.next_node]\n",
    "                    del self.nodes[edge.next_node]\n",
    "\n",
    "    def _print_connections(self):\n",
    "        # count = 0\n",
    "        # for node in self.nodes:\n",
    "        #     if len(self.connections[node]) == 0:\n",
    "        #         count += 1\n",
    "        # print(count)\n",
    "        print(self.connections)\n",
    "\n",
    "    def _print_nodes(self):\n",
    "        for node in self.nodes:\n",
    "            print(node)\n",
    "            print(self.nodes[node].in_degree)\n",
    "            print(self.nodes[node].out_degree)\n",
    "\n",
    "    def _find_euler_path(self, v):\n",
    "        for edge in self.connections_temp[v]:\n",
    "            if edge.coverage:\n",
    "                edge.coverage -= 1\n",
    "                self.string += edge.string[(self.k-1):]\n",
    "                self._find_euler_path(edge.next_node)\n",
    "\n",
    "    def assemble(self, multiple_start=False):\n",
    "        in_degrees = [self.nodes[i].in_degree for i in self.connections.keys()]\n",
    "        nodes_min_in_degrees = [list(self.connections.keys())[i] for i in range(len(in_degrees)) if in_degrees[i] == min(in_degrees)]\n",
    "        if multiple_start:\n",
    "            for start in nodes_min_in_degrees:\n",
    "                self.connections_temp = deepcopy(self.connections)\n",
    "                self.string = self.connections[start][0].string[:(self.k-1)]\n",
    "                self._find_euler_path(start)\n",
    "        else:\n",
    "            start = nodes_min_in_degrees[0]\n",
    "            self.connections_temp = deepcopy(self.connections)\n",
    "            self.string = self.connections[start][0].string[:(self.k-1)]\n",
    "            self._find_euler_path(start)\n",
    "        return self.string\n",
    "    \n",
    "    def visualize(self, interactive=False, show_edges=False, layout=nx.kamada_kawai_layout):\n",
    "        G = nx.DiGraph() \n",
    "        for node in self.connections:\n",
    "            for edge in self.connections[node]:\n",
    "                G.add_edge(edge.prev_node, edge.next_node, string=edge.string)\n",
    "        if interactive:\n",
    "            nt = Network('2000px', '2000px', directed=True)\n",
    "            nt.from_nx(G)\n",
    "            nt.show('test.html', notebook=False)\n",
    "        else:\n",
    "            for node in self.nodes:\n",
    "                G.add_node(node)\n",
    "            pos = layout(G)\n",
    "            plt.figure(figsize=(15,15))\n",
    "            nx.draw_networkx(G, pos, font_size=5)\n",
    "            if show_edges:\n",
    "                nx.draw_networkx_edge_labels(G, pos, font_size=8)\n",
    "            plt.show()\n",
    "\n",
    "    def get_statistics(self):\n",
    "        edge_coverage = []\n",
    "        node_degrees = []\n",
    "        n_edges = 0\n",
    "        n_nodes = len(self.nodes)\n",
    "        for node in self.connections:\n",
    "            node_degrees.append(self.nodes[node].in_degree + self.nodes[node].out_degree)\n",
    "            for edge in self.connections[node]:\n",
    "                edge_coverage.append(edge.coverage)\n",
    "                n_edges += 1\n",
    "        print('Number edges:', n_edges)\n",
    "        print('Number nodes:', n_nodes)\n",
    "        figure, axis = plt.subplots(1, 2, figsize=(25, 5))\n",
    "        counts_edge, bins_edge = np.histogram(edge_coverage)\n",
    "        axis[0].hist(bins_edge[:-1], 10, weights=counts_edge)\n",
    "        axis[0].set_title('Edge coverage')\n",
    "        # axis[0].set_xticks(np.arange(1, max(edge_coverage)+1, len(edge_coverage)//10))\n",
    "        counts_node, bins_node = np.histogram(node_degrees)\n",
    "        axis[1].hist(bins_node[:-1], 10, weights=counts_node)\n",
    "        axis[1].set_title('Node degrees')\n",
    "        # axis[1].set_xticks(np.arange(1, max(node_degrees)+1, len(node_degrees)//10))\n",
    "        plt.show()\n",
    "    \n",
    "    def get_shared_motif(self, true_seq):\n",
    "        if not self.string:\n",
    "            print('Do assemble first!')\n",
    "            return\n",
    "        s_seq = self.string if len(self.string) <= len(true_seq) else true_seq\n",
    "        l_seq = self.string if len(self.string) > len(true_seq) else true_seq\n",
    "        motif = set()\n",
    "        for i in range(len(s_seq)):\n",
    "            for j in range(i+1, len(s_seq)+1):\n",
    "                motif.add(s_seq[i:j])\n",
    "        update_motif = list(motif)\n",
    "        for m in update_motif:\n",
    "            if m not in l_seq:\n",
    "                motif.remove(m)\n",
    "        motif = list(motif)\n",
    "        longest_motif = motif[np.argmax([len(i) for i in motif])]\n",
    "        print('Longest shared motif is:\\n', longest_motif, end='\\n\\n\\n')\n",
    "\n",
    "        \n",
    "\n",
    "true_seq_1 = 'AGATGAATGGACCGGCCATATAAGTAAG'\n",
    "test_sequences = ['CCGTTG', 'CGTTAC', 'GTTGCA', 'TGCAGG']\n",
    "\n",
    "\n",
    "G = DBGraph([true_seq], 6)\n",
    "G.gen_kmers()\n",
    "G.create_graph()\n",
    "s1 = true_seq\n",
    "s2 = G.assemble()\n",
    "# G._print_connections()\n",
    "# G._print_connections()\n",
    "\n",
    "# G.visualize(show_edges=False, layout=nx.kamada_kawai_layout)\n",
    "# G.delete_tails()\n",
    "# print(s1)\n",
    "# print(s2)\n",
    "# G.get_shared_motif(s1)\n",
    "# print()\n",
    "# G.visualize(show_edges=False, layout=nx.kamada_kawai_layout)\n",
    "# G.get_statistics()\n",
    "# G.visualize(show_edges=False, layout=nx.kamada_kawai_layout)\n",
    "\n",
    "# G.compact()\n",
    "# G._print_nodes()\n",
    "# G.visualize(show_edges=False, layout=nx.kamada_kawai_layout)\n",
    "# G.get_shared_motif(s1)\n",
    "# G.get_statistics()\n",
    "# G._print_connections()\n",
    "\n",
    "# G.delete_tails()\n",
    "# G.compact()\n",
    "G.visualize(interactive=True, show_edges=False, layout=nx.kamada_kawai_layout)\n",
    "# G._print_nodes()\n",
    "\n",
    "\n",
    "# G.compact()\n",
    "\n",
    "\n",
    "# # s1 = true_seq\n",
    "# # s2 = G.assemble()\n",
    "# print(s1)\n",
    "# print(s2)\n",
    "# # print()\n",
    "# G.visualize(show_edges=False, layout=nx.kamada_kawai_layout)\n",
    "# G._print_connections()\n",
    "# G.get_statistics()\n",
    "# G.get_shared_motif(s1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Сжатие графа  \n",
    "Научитесь производить сжатие графа Де Брюина. При сжатии не забывайте склеивать подстроки на ребрах и обновлять покрытие ребер. Пересчитывайте покрытие склеиваемых ребер как взвешенное среднее, где вес соответствует длине подстроки, соответствующей ребру."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
