{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\Python\\envs\\.general\\Lib\\site-packages\\Bio\\pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import Bio\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import substitution_matrices\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read sequence from file:\n",
      "ID: NC_000023\n",
      "Description: NC_000023\n",
      "Sequence: AAGTGCTTGGGGCGGCGAGCATGGCGGCAGCGGCTGCAGGCCTGGGCGGCGGCGGCGCCGGCCCGGGACCCGAGGCCGGGGACTTCCTGGCCCGCTACCGGCTGGTATCGAACAAGCTGAAGAAGCGGTTCCTGCGGAAGCCGAACGTGGCGGAGGCCGGCGAGCAGTTCGGACAGCTGGGCCGGGAGCTGCGCGCCCAGGAGTGTCTGCCCTACGCGGCCTGGTGCCAGCTGGCGGTGGCGCGCTGCCAGCAGGCGCTCTTCCACGGGCCCGGGGAGGCGCTGGCCCTCACCGAGGCCGCCCGCCTCTTCCTGCGGCAGGAGCGCGACGCGCGCCAGCGCCTGGTCTGCCCCGCCGCCTACGGGGAGCCGCTGCAGGCCGCCGCCAGCGCCCTGGGCGCCGCGGTGCGTCTGCACCTCGAGCTGGGCCAGCCGGCCGCCGCCGCCGCCCTCTGCCTCGAGCTGGCCGCCGCCCTGCGCGACCTGGGCCAGCCGGCCGCCGCCGCCGGTCACTTCCAGCGCGCCGCCCAGCTCCAGCTGCCCCAGCTGCCCCTGGCCGCGCTGCAGGCGCTTGGCGAGGCCGCCTCCTGCCAGCTGCTGGCGCGCGACTACACCGGCGCCCTGGCGGTCTTCACGCGCATGCAGCGCCTGGCGCGGGAGCACGGCAGCCACCCGGTGCAGTCACTGCCGCCGCCCCCGCCGCCGGCACCCCAGCCCGGGCCCGGGGCGACGCCCGCCCTACCGGCCGCGCTGCTTCCTCCGAACTCCGGCTCGGCGGCGCCCTCTCCCGCCGCCCTGGGCGCCTTCTCGGACGTGCTGGTCCGCTGCGAGGTGTCCCGCGTGCTGCTGCTGCTCCTCCTGCAACCACCGCCCGCCAAGCTGCTGCCGGAGCACGCCCAGACCCTGGAGAAGTACTCCTGGGAGGCTTTTGACAGCCACGGGCAGGAGAGCAGCGGCCAGCTTCCCGAGGAGCTCTTTCTGCTGCTCCAGTCTTTGGTCATGGCTACCCACGAAAAGGACACGGAAGCCATCAAGTCGCTGCAGGTGGAGATGTGGCCACTGTTGACTGCTGAGCAGAACCACCTCCTTCACCTCGTTCTGCAAGAAACCATCTCCCCCTCAGGACAGGGAGTCTGATCCATCCCATTCACCCAGTGACTTCTTTTTGCCCAGGCCTGGACTTTTTGCATCAGTCACGTTAACCAGATGACTTTGCCTGTTACCAAACCTCATGCATCCACGTTTGCGTCTGGGGAGGAATAAAAAGACATCGTTCCCGCTTCTGCGTTTTGTTATTCCTACTGCCGCCATAGGAATTATTTCGTTGGCTGAACGTTACCAGCACCCCGAGAACACATTTTGATAGAATCAGAGTAGAGGACATGGCTGTCTTCTAAAAAGCCACGACATGAAAATGACAATCCCTTTCGTCTCCTTCCTCCGCTGCTTCCACCTAACGCAGCCTCCTGCCTCCGCCTTTGTTTCATAGTGAGGATTTTATTTTGCACGGCGCCCTCCCTCTAAATACCTACCCTAGATGATTTCATCCTGCCCCTCACTTCTTAACATATTCCTGTGTCTTTGTAATGGCCAAATTTCTCCTTCCACTCGTCCGCACGGTACGTCTTCATGGGAGTCATTTTATTCCTTACAGCTTCCTTGTAACCACAGCCCTAAATCCATGATAAAGCTACTCTTGTACTCCTCA\n",
      "Read sequence from file:\n",
      "ID: NC_000023_random\n",
      "Description: NC_000023_random\n",
      "Sequence: AAGTGCTTGGGGCGGCGAGCATGGCGGCAGCGGCTGCAGGCCTGGGCGGCGGCGGCGCCGGCCCGGGACCCGAGGCCGGGGACTTCCTGGCCCGCTACCGGCTGGTATCGAACAAGCTGAAGAAGCGGTTCCTGCGGAAGCCGAACGTGGCGGAGGCCGGCGAGCAGTTCGGACAGCTGGGCCGGGAGCTGCGCGCCCAGGAGTGTCTGCCCTACGCGGCCTGGATTTTTTTGCCAGCTGGCGGTGGCGCGCTGCCAGCAGGCGGGGGGCCTCTTCCACGGGCCCGGGGAGGCGCTGGCCCTCACCGAGGCCGCCCGCCTCTTCCTGCGGCAGGAGCGCGACGCGCGCCAGCGCCTGGTCTGCCCCGCCGCCTACGGGGAGCCGCTGCAGGCCGCCGCCAGCGCCCTGGGCGCCGCGGTGCGTCTGCACCTCGAGCTGGGCCAGCCGGCCGCCGCCGCCGCCCTCTGCCTCGAGCTGGCCGCCGCCCTGCGCGACCTGGGCCAGCCGGCCGCCGCCGCCGGTCACTTCCAGCGCGCCGCCCAGCTCCAGCTGCCCCAGCTGCCCCTGGCCGCGCTGCAGGCGCTTGGCGAGGCCGCCTCCTGCCAGCTGCTGGCGCGCGACTACACCGGCGCCCTGGCGGTCTTCACGCGCATGCAGCGCCTGGCGCGGGAGCACAAATTGGCAGCCACCCGGTGCAGTCACTGCCGCCGCCCCCGCCGCCGGCACCCCAGCCCGGGCCCGGGGCGACGCCCGCCCTACCGGCCGCGCTGTCGGCGGCGCCCTCTCCCGCCGCCCTGGGCGCCTTCTCGGACGTGCTGGTCCGCTGCGAGGTGTCCCGCGTGCTGCTGCTGCTCCTCCTGCAACCACCGCCCGCCAAGCTGCTGCCGGAGCACGCCCAGACCCTGGAGAAGTACTCCTGGGAGGCTTTTGACAGCCACGGGCAGGAGAGCAGCGGCCAGCTTCCCGAGGAGCTCTTTCTGCTGCTCCAGTCTTTGGTCATGGCTACCCACGAAAAGGACACGGAAGCCATCAAGTCGCTGCAGGTGGAGATGTGGCCACTGTTGACTGCTGAGCAGGGTTTCCGAACCACCTCCTTCACCTCGTTCTGCAAGAAACCATCTCCCCCTCAGGACAGGGAGTCTGATCCATCCCATTCACCCAGTGACTTCTTTTTGCCCAGGCCTGGACTTAGTCACGTTAACCAGATGACTTTGCCTGTTACCAAACCTCATGCATCCACGTTTGCGTCTGGGGAGGAATAAAAAGACATCGTTCCCGCTTCTGCGTTTTGTTATTCCTACTGCCGCCATAGGAATTATTTCGTTGGCTGAACGTTACCAGCACCCCGAGAACACATTTTGATAGAATCAGAGTAGAGGACATGGCTGTCTTCTAAAAAGCCACGACATGAAAATGACAATCCCTTTCGTCTCCTTCCAAAAAAACCCTCCCCCGCTGCTTCCACCTAACGCAGCCTCCTGCCTCCGCCTTTGTTTCATAGTGAGGATTTTATTTTGCACGGCGCCCTCCCTCTAAATACCTACCCTAGATGTTTCATCCTGCCCCTCACTTCTTAACATATTCCTGTGTCTTTGTAATGGCCAAATTTCTCCTTCACTCGTCCGCACGGTACGTCTTCATGGGAGTCATTTTATTCCTTACAGCTTCCTTGTAACCACAGCCCTAAATCCATATAAAGCTACTCTTTACTCCTCA\n"
     ]
    }
   ],
   "source": [
    "fasta_sequences = SeqIO.parse(r\"D:\\HSE\\BioinformaticsCourse2024\\homework\\1_1\\data\\f8.fasta\", \"fasta\")\n",
    "f8_sequences = []\n",
    "for record in fasta_sequences.records:\n",
    "    print(\"Read sequence from file:\")\n",
    "    print(\"ID:\", record.id)\n",
    "    print(\"Description:\", record.description)\n",
    "    print(\"Sequence:\", record.seq)\n",
    "    f8_sequences.append(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read sequence from file:\n",
      "ID: first\n",
      "Description: first\n",
      "Sequence: GATTACA\n",
      "Read sequence from file:\n",
      "ID: second\n",
      "Description: second\n",
      "Sequence: CATTAGA\n"
     ]
    }
   ],
   "source": [
    "fasta_sequences = SeqIO.parse(r\"D:\\HSE\\BioinformaticsCourse2024\\homework\\1_1\\data\\gattaca.fasta\", \"fasta\")\n",
    "gattaca_sequences = []\n",
    "for record in fasta_sequences.records:\n",
    "    print(\"Read sequence from file:\")\n",
    "    print(\"ID:\", record.id)\n",
    "    print(\"Description:\", record.description)\n",
    "    print(\"Sequence:\", record.seq)\n",
    "    gattaca_sequences.append(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_seq(len1, len2):\n",
    "    \n",
    "    '''Generates absolute random sequence of a given length'''\n",
    "    seq1 = ''\n",
    "    seq2 = ''\n",
    "    for _ in range(len1):\n",
    "        seq1 += random.choice('ARNDCQEGHILKMFPSTWYVBZX')\n",
    "    for _ in range(len2):\n",
    "        seq2 += random.choice('ARNDCQEGHILKMFPSTWYVBZX')\n",
    "    return seq1, seq2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Нидлман Вунш"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Реализуйте алгоритм Нидлмана Вунша для выравнивания последовательностеей. На вход принимается две строки, матрица замен и стоимость гэпа. В результате верните оптимальное выравнивание и его вес. При проверке помните, что оптимальных выравниваний может быть несколько, но вес у них должен совпадать."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "My solution:\n",
      "Score = 115.0:\n",
      "target\t\t  0 YAFDLGYTCMFP-VLLGGGELHIV-QKETYTAPDEIAHYIKEHGITYIKLTPSLFHTIVNTASFAFDANFE-S-LRLIVLGGEKIIPIDVIAFRKMYGHT-EFINHYGPTEATIGA 111\n",
      "\t\t  0 -|||..-...|.-.||.||.|-||-..|....|......||...||....||.|---..............-|-|.....|.......|........|.|-...|.||.|||.|.. 116\n",
      "query\t\t  0 -AFDVS-AGDFARALLTGGQL-IVCPNEVKMDPASLYAIIKKYDITIFEATPAL---VIPLMEYIYEQKLDISQLQILIVGSDSCSMEDFKTLVSRFGSTIRIVNSYGVTEACIDS 110\n",
      "\n",
      "Test solution:\n",
      "Score = 115.0:\n",
      "target            0 YAFDLGYTCMFP-VLLGGGELHIV-QKETYTAPDEI-AHYIKEHGITYIKLTPSLFHTIV\n",
      "                  0 -|||..-...|.-.||.||.|-||-..|....|...-|-.||...||....||.|---..\n",
      "query             0 -AFDVS-AGDFARALLTGGQL-IVCPNEVKMDPASLYA-IIKKYDITIFEATPAL---VI\n",
      "\n",
      "target           57 NTASFAFDANFE-S-LRLIVLGGEKIIPIDVIAFRKMYGHT-EFINHYGPTEATIGA 111\n",
      "                 60 ............-|-|.....|.......|........|.|-...|.||.|||.|.. 117\n",
      "query            53 PLMEYIYEQKLDISQLQILIVGSDSCSMEDFKTLVSRFGSTIRIVNSYGVTEACIDS 110\n",
      "\n",
      "Score = 115.0:\n",
      "target            0 YAFDLGYTCMFP-VLLGGGELHIV-QKETYTAPDEIAHYIKEHGITYIKLTPSLFHTIVN\n",
      "                  0 -|||..-...|.-.||.||.|-||-..|....|......||...||....||.|---...\n",
      "query             0 -AFDVS-AGDFARALLTGGQL-IVCPNEVKMDPASLYAIIKKYDITIFEATPAL---VIP\n",
      "\n",
      "target           58 TASFAFDANFE-S-LRLIVLGGEKIIPIDVIAFRKMYGHT-EFINHYGPTEATIGA 111\n",
      "                 60 ...........-|-|.....|.......|........|.|-...|.||.|||.|.. 116\n",
      "query            54 LMEYIYEQKLDISQLQILIVGSDSCSMEDFKTLVSRFGSTIRIVNSYGVTEACIDS 110\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def NeedlemanWunsch(seq1, seq2, subsistution_matrix, gap_penalty):\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    dp = [[0 for _ in range(m+1)] for _ in range(n+1)]\n",
    "    for i in range(n+1):\n",
    "        dp[i][0] = i*gap_penalty\n",
    "    for i in range(m+1):\n",
    "        dp[0][i] = i*gap_penalty\n",
    "    for i in range(1, n+1):\n",
    "        for j in range(1, m+1):\n",
    "            aminoacid1_idx = subsistution_matrix.alphabet.find(seq1[i-1])\n",
    "            aminoacid2_idx = subsistution_matrix.alphabet.find(seq2[j-1])\n",
    "            dp[i][j] = max(\n",
    "                dp[i-1][j-1] + subsistution_matrix[aminoacid1_idx][aminoacid2_idx],\n",
    "                dp[i-1][j] + gap_penalty,\n",
    "                dp[i][j-1] + gap_penalty,\n",
    "            )\n",
    "    score = dp[n][m]\n",
    "\n",
    "    alignment1 = ''\n",
    "    alignment2 = ''\n",
    "    while n > 0 or m > 0:\n",
    "        aminoacid1_idx = subsistution_matrix.alphabet.find(seq1[n-1])\n",
    "        aminoacid2_idx = subsistution_matrix.alphabet.find(seq2[m-1])\n",
    "        if n > 0 and m > 0 and dp[n][m] == dp[n-1][m-1] + subsistution_matrix[aminoacid1_idx][aminoacid2_idx]:\n",
    "            alignment1 = seq1[n-1] + alignment1\n",
    "            alignment2 = seq2[m-1] + alignment2\n",
    "            n -= 1\n",
    "            m -= 1\n",
    "        elif n > 0 and dp[n][m] == dp[n-1][m] + gap_penalty:\n",
    "            alignment1 = seq1[n-1] + alignment1\n",
    "            alignment2 = '-' + alignment2\n",
    "            n -= 1\n",
    "        else:\n",
    "            alignment2 = seq2[m-1] + alignment2\n",
    "            alignment1 = '-' + alignment1\n",
    "            m -= 1\n",
    "    return (score, alignment1, alignment2)\n",
    "\n",
    "\n",
    "matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "\n",
    "#  ARNDCQEGHILKMFPSTWYVBZX\n",
    "# seq1, seq2 = gattaca_sequences\n",
    "seq1 = 'YAFDLGYTCMFPVLLGGGELHIVQKETYTAPDEIAHYIKEHGITYIKLTPSLFHTIVNTASFAFDANFESLRLIVLGGEKIIPIDVIAFRKMYGHTEFINHYGPTEATIGA'\n",
    "seq2 = 'AFDVSAGDFARALLTGGQLIVCPNEVKMDPASLYAIIKKYDITIFEATPALVIPLMEYIYEQKLDISQLQILIVGSDSCSMEDFKTLVSRFGSTIRIVNSYGVTEACIDS'\n",
    "# seq1, seq2 = generate_seq(10, 20)  # absolute random sequences\n",
    "\n",
    "my_score, my_alignment1, my_alignment2 = NeedlemanWunsch(seq1, seq2, matrix, -4)\n",
    "comparison_str = ''\n",
    "for i in range(len(my_alignment1)):\n",
    "    if my_alignment1[i] == my_alignment2[i]:\n",
    "        comparison_str += '|'\n",
    "    elif my_alignment1[i] != my_alignment2[i] and my_alignment1[i] != '-' and my_alignment2[i] != '-':\n",
    "        comparison_str += '.'\n",
    "    else:\n",
    "        comparison_str += '-'\n",
    "\n",
    "print(f'My solution:\\nScore = {my_score}:')\n",
    "print(f'target\\t\\t  0 {my_alignment1} {len(seq1)}')\n",
    "print(f'\\t\\t  0 {comparison_str} {len(comparison_str)}')\n",
    "print(f'query\\t\\t  0 {my_alignment2} {len(seq2)}\\n')\n",
    "\n",
    "aligner = Bio.Align.PairwiseAligner()\n",
    "aligner.substitution_matrix = matrix\n",
    "aligner.open_gap_score = -4\n",
    "aligner.extend_gap_score = -4\n",
    "print('Test solution:')\n",
    "for alignment in aligner.align(seq1, seq2):\n",
    "    print(\"Score = %.1f:\" % alignment.score)\n",
    "    print(alignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Афинные гэпы"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Реализуйте выравнивание с афинными гэпами, алгоритм на вход принимает две строки, матрицу замен, штраф за начало гэпа α, и за его продолжение β. В результате возвращает выравнивание и его вес. Сложность алгоритма квадратичная по памяти и по времени."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test solution:\n",
      "Score = 8.0:\n",
      "target            0 YAFDL--GYTC  9\n",
      "                  0 -|||.--|... 11\n",
      "query             0 -AFDVSAGDFA 10\n",
      "\n",
      "Score = 8.0:\n",
      "target            0 YAFDL--G-YTC  9\n",
      "                  0 -|||.--|-.-. 12\n",
      "query             0 -AFDVSAGDF-A 10\n",
      "\n",
      "Score = 8.0:\n",
      "target            0 YAFDL--G-YTC  9\n",
      "                  0 -|||.--|-..- 12\n",
      "query             0 -AFDVSAGDFA- 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def AfinityGaps(seq1, seq2, subsistution_matrix, open_gap_penalty, extend_gap_penalty):\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    dp = [[0 for _ in range(m+1)] for _ in range(n+1)]\n",
    "    dp_insertion = [[0 for _ in range(m+1)] for _ in range(n+1)]\n",
    "    dp_deletion = [[0 for _ in range(m+1)] for _ in range(n+1)]\n",
    "    for i in range(1, m+1):\n",
    "        dp_insertion[0][i] = -10e6\n",
    "    for i in range(1, n+1):\n",
    "        dp_deletion[i][0] = -10e6\n",
    "    for i in range(1, n+1):\n",
    "        dp[i][0] = open_gap_penalty + (i-1)*extend_gap_penalty\n",
    "    for i in range(1, m+1):\n",
    "        dp[0][i] = open_gap_penalty + (i-1)*extend_gap_penalty\n",
    "    for i in range(1, n+1):\n",
    "        for j in range(1, m+1):\n",
    "            dp_insertion[i][j] = max(\n",
    "                dp_insertion[i-1][j] + extend_gap_penalty,\n",
    "                dp[i-1][j] + open_gap_penalty\n",
    "            )\n",
    "            dp_deletion[i][j] = max(\n",
    "                dp_deletion[i][j-1] + extend_gap_penalty,\n",
    "                dp[i][j-1] + open_gap_penalty\n",
    "            )\n",
    "            aminoacid1_idx = subsistution_matrix.alphabet.find(seq1[i-1])\n",
    "            aminoacid2_idx = subsistution_matrix.alphabet.find(seq2[j-1])\n",
    "            dp[i][j] = max(\n",
    "                dp[i-1][j-1] + subsistution_matrix[aminoacid1_idx][aminoacid2_idx],\n",
    "                dp_insertion[i][j],\n",
    "                dp_deletion[i][j],\n",
    "            )\n",
    "    score = dp[n][m]\n",
    "\n",
    "    alignment1 = ''\n",
    "    alignment2 = ''\n",
    "    current_matrix = 'substitution'\n",
    "    while n > 0 or m > 0:\n",
    "        if n == 0:\n",
    "            current_matrix = 'deletion'\n",
    "        elif m == 0:\n",
    "            current_matrix = 'insertion'\n",
    "        aminoacid1_idx = subsistution_matrix.alphabet.find(seq1[n-1])\n",
    "        aminoacid2_idx = subsistution_matrix.alphabet.find(seq2[m-1])\n",
    "        if current_matrix == 'substitution':\n",
    "            if dp[n][m] == dp_insertion[n][m]:\n",
    "                current_matrix == 'insertion'\n",
    "            elif dp[n][m] == dp_deletion[n][m]:\n",
    "                current_matrix = 'deletion'\n",
    "            elif dp[n][m] == dp[n-1][m-1] + subsistution_matrix[aminoacid1_idx][aminoacid2_idx]:\n",
    "                alignment1 = seq1[n-1] + alignment1\n",
    "                alignment2 = seq2[m-1] + alignment2\n",
    "                n -= 1\n",
    "                m -= 1\n",
    "        elif current_matrix == 'insertion':\n",
    "            if dp_insertion[n][m] == dp[n-1][m] + open_gap_penalty:\n",
    "                current_matrix = 'substitution'\n",
    "            alignment1 = seq1[n-1] + alignment1\n",
    "            alignment2 = '-' + alignment2\n",
    "            n -= 1\n",
    "        elif current_matrix == 'deletion':\n",
    "            if dp_deletion[n][m] == dp[n][m-1] + open_gap_penalty:\n",
    "                current_matrix = 'substitution'\n",
    "            alignment2 = seq2[m-1] + alignment2\n",
    "            alignment1 = '-' + alignment1\n",
    "            m -= 1\n",
    "    return (score, alignment1, alignment2)\n",
    "\n",
    "\n",
    "matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "\n",
    "#  ARNDCQEGHILKMFPSTWYVBZX\n",
    "# seq1, seq2 = gattaca_sequences\n",
    "seq1 = 'YAFDLGYTC'\n",
    "seq2 = 'AFDVSAGDFA'\n",
    "# seq1, seq2 = generate_seq(10, 20)  # absolute random sequences\n",
    "\n",
    "open_gap_score = -4\n",
    "extend_gap_score = -2\n",
    "\n",
    "my_score, my_alignment1, my_alignment2 = AfinityGaps(seq1, seq2, matrix, open_gap_score, extend_gap_score)\n",
    "comparison_str = ''\n",
    "for i in range(len(my_alignment1)):\n",
    "    if my_alignment1[i] == my_alignment2[i]:\n",
    "        comparison_str += '|'\n",
    "    elif my_alignment1[i] != my_alignment2[i] and my_alignment1[i] != '-' and my_alignment2[i] != '-':\n",
    "        comparison_str += '.'\n",
    "    else:\n",
    "        comparison_str += '-'\n",
    "\n",
    "print(f'My solution:\\nScore = {my_score}:')\n",
    "print(f'target\\t\\t  0 {my_alignment1} {len(seq1)}')\n",
    "print(f'\\t\\t  0 {comparison_str} {len(comparison_str)}')\n",
    "print(f'query\\t\\t  0 {my_alignment2} {len(seq2)}\\n')\n",
    "\n",
    "aligner = Bio.Align.PairwiseAligner()\n",
    "aligner.substitution_matrix = matrix\n",
    "aligner.open_gap_score = open_gap_score\n",
    "aligner.extend_gap_score = extend_gap_score\n",
    "print('Test solution:')\n",
    "for alignment in aligner.align(seq1, seq2):\n",
    "    print(\"Score = %.1f:\" % alignment.score)\n",
    "    print(alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def needleman_wunsch2(A, B, kappa = 1, mu = -1, delta = -1, epsilon = -1):\n",
    "    \"\"\"\n",
    "        Needleman-Wunsch algorithm for sequience alignment with affine gap penalty (Gotoh method).\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        A, B : str\n",
    "            Two sequiences\n",
    "        kappa : num, optional \n",
    "            Match reward (default is +1)\n",
    "        mu : num, optional\n",
    "            Mismatch penalty (default is -1)\n",
    "        delta : num, optional\n",
    "            Gap open penalty (default is -1)\n",
    "        epsilon : num, optional\n",
    "            Gap extend penalty (default is -1)\n",
    "        \"\"\"\n",
    "\n",
    "    n = len(A)\n",
    "    m = len(B)\n",
    "\n",
    "    # Best score table\n",
    "    D = [[delta + (i-1)*epsilon for i in range(n+1)] if j == 0 else [delta + (j-1)*epsilon] + [0]*n for j in range(m+1)]\n",
    "    D[0][0] = 0\n",
    "    # Score if x_i aligns to a gap after y_j\n",
    "    P = [[-1000] + [0] * (n) for _ in range(m+1)]\n",
    "    P[0][0] = 0\n",
    "    # Score if y_j aligns to a gap after x_i\n",
    "    Q = [([-1000] * (n+1) if j == 0 else [0] * (n+1)) for j in range(m+1)]\n",
    "    Q[0][0] = 0\n",
    " \n",
    "    # Filling the table\n",
    "    for i in range(1, n+1):\n",
    "        for j in range(1, m+1):\n",
    "            P[j][i] = max(D[j][i-1]+delta, P[j][i-1]+epsilon)\n",
    "            Q[j][i] = max(D[j-1][i]+delta, Q[j-1][i]+epsilon)\n",
    "            D[j][i] = max(D[j-1][i-1] + (kappa if A[i-1] == B[j-1] else mu), \n",
    "                        P[j][i], \n",
    "                        Q[j][i])\n",
    "     # Tracing path back to origin\n",
    "    i = n\n",
    "    j = m\n",
    "    A_aligned = ''\n",
    "    B_aligned = ''\n",
    "    table = 'D'\n",
    "\n",
    "    while i > 0 or j > 0:\n",
    "        if i == 0:\n",
    "            table = 'Q'\n",
    "        elif j == 0:\n",
    "            table = 'P'\n",
    "\n",
    "        if table == 'D':\n",
    "            if D[j][i] == Q[j][i]:\n",
    "                # Switch table\n",
    "                table = 'Q'\n",
    "            elif D[j][i] == P[j][i]:\n",
    "                # Switch table\n",
    "                table = 'P'\n",
    "            elif D[j][i] == D[j-1][i-1] + (kappa if A[i-1] == B[j-1] else mu):\n",
    "                # Came from left-top \n",
    "                A_aligned = A[i-1] + A_aligned\n",
    "                B_aligned = B[j-1] + B_aligned\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            \n",
    "        elif table == 'P':\n",
    "            if P[j][i] == D[j][i-1] + delta:\n",
    "                table = 'D'\n",
    "            A_aligned = A[i-1] + A_aligned\n",
    "            B_aligned = '_' + B_aligned\n",
    "            i -= 1\n",
    "        elif table == 'Q':\n",
    "            if Q[j][i] == D[j-1][i] + delta:\n",
    "                table = 'D'\n",
    "            A_aligned = '_' + A_aligned\n",
    "            B_aligned = B[j-1] + B_aligned\n",
    "            j -= 1\n",
    "\n",
    "    # Return two aligned sequiences and total score\n",
    "    return A_aligned, B_aligned, D[m][n]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def affine_gap_alignment(seq1, seq2, weight_match, weight_mismatch, open_gap_penalty, continue_gap_penalty):\n",
    "    a, b, t = needleman_wunsch2(seq1, seq2, kappa = weight_match, mu = weight_mismatch, delta = open_gap_penalty, epsilon = continue_gap_penalty)\n",
    "    print(a)\n",
    "    print(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YAFDL__GYTC\n",
      "_AFDVSAGDFA\n"
     ]
    }
   ],
   "source": [
    "affine_gap_alignment(seq1, seq2, 1, -1, -4, -2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".general",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
